### ISSUE TITLE
[RFC] Activation Checkpoint API improvements

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#70060]
with @zhaojuanmao 

Activation checkpointing is an important technique to trade off compute for more GPU memory, thereby enabling larger batch sizes or the training of larger models.

It is especially relevant in the case of large model training with `FullyShardedDataParallel`, where the goal is to train models that are much bigger than what can fit in a single GPU by using a variety of techniques: sharding, activation offloading, parameter and gradient offloading. 

As a result, we would like to add the following APIs to `torch.utils.checkpoint`:

1. A `checkpoint_wrapper` function that wraps an `nn.Module` instance so that it is always run with checkpointing. This is helpful so that the user will not have to call `checkpoint` everytime the model's forward pass is run in code. It also allows for better composition with FSDP and DDP, i.e. the user can choose whether to do `FSDP(checkpoint_wrapper(model))` or `checkpoint_wrapper(FSDP(model))`, which, depending on model and other training conditions can have different performance implications. A prototype of `checkpoint_wrapper` has been added with this PR: https://github.com/pytorch/pytorch/pull/69955, and we would like to upstream this to `torch.utils.checkpoint`. The implementation is derived from [fairscale's checkpoint_wrapper](https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/checkpoint/checkpoint_utils.py).

2. Support for _activation offloading_. To understand this, assume that we have the following model structure: `nn.Sequential(checkpoint_wrapper(l1), checkpoint_wrapper(l2), ..., checkpoint_wrapper(ln))`. In this case, _inner_ activations are within each checkpointed layer and memory is saved by discarding and recomputing them. Although there are still so-called _outer_ activations between the layers that need to be saved on GPU for recomputation. _activation offloading_ refers to the idea of offloading these activations to CPU to additionally save on GPU memory. It is especially helpful when outer activations are very large. A prototype for activation offloading is here: https://github.com/pytorch/pytorch/pull/70024. The implementation is derived from [fairscale's checkpoint_wrapper](https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/checkpoint/checkpoint_utils.py)

In addition, there are 2 checkpoint-based implementations, the traditional reentrant-backwards passed one and the recently added (https://github.com/pytorch/pytorch/pull/69508) checkpoint based on saved tensor hooks. We would like to support both APIs for the time being, because reentrant-based checkpoint is still the default offering and used by FSDP, DDP as well, although we will migrate to saved tensor hooks based checkpointing in the future.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang
